### PR TITLE
Fix #2992 of Creating new solution using "dotnet new sln" in .NET 5.0.201 SDK has wrong/invalid solution version number

### LINKED ISSUE #2992: Creating new solution using "dotnet new sln" in .NET 5.0.201 SDK has wrong/invalid solution version number
## Problem

When we create new solution using "dotnet new sln" in .NET 5.0.201 SDK has wrong/invalid solution version number in the `VisualStudioVersion`. This version is invalid, because there's no Visual Studio solution has higher version than "16.6.*.*" at this moment (at least in observed 16.9.x)

Reason: this version cannot be used to enforce minimum version of Visual Studio to open. 

## Steps to reproduce

The steps are:

1. Create new solution using `dotnet new sln`. Example: `dotnet new sln -n SolutionNET5`
2. Open the solution file in notepad other text editor
3. The version of the VisualStudioVersion is `16.6.30114.105`
4. Update the MinimumVisualStudioVersion to be the same as VisualStudioVersion
5. Open the solution in Visual Studio 2019 16.9.x (error happened in 16.9.0, 16.9.1, 16.9.2)

Full contents after `dotnet new sln` is:
```
Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio Version 16
VisualStudioVersion = 16.6.30114.105
MinimumVisualStudioVersion = 10.0.40219.1
Global
        GlobalSection(SolutionConfigurationPlatforms) = preSolution
                Debug|Any CPU = Debug|Any CPU
                Debug|x64 = Debug|x64
                Debug|x86 = Debug|x86
                Release|Any CPU = Release|Any CPU
                Release|x64 = Release|x64
                Release|x86 = Release|x86
        EndGlobalSection
        GlobalSection(SolutionProperties) = preSolution
                HideSolutionNode = FALSE
        EndGlobalSection
EndGlobal
```

The error after I update the MinimumVisualStudioVersion value to `16.6.30114.105` is:

image

If I use .NET Core 3.1.406 SDK to create new solution, I have this:

```
Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio 15
VisualStudioVersion = 15.0.26124.0
MinimumVisualStudioVersion = 15.0.26124.0
Global
        GlobalSection(SolutionConfigurationPlatforms) = preSolution
                Debug|Any CPU = Debug|Any CPU
                Debug|x64 = Debug|x64
                Debug|x86 = Debug|x86
                Release|Any CPU = Release|Any CPU
                Release|x64 = Release|x64
                Release|x86 = Release|x86
        EndGlobalSection
        GlobalSection(SolutionProperties) = preSolution
                HideSolutionNode = FALSE
        EndGlobalSection
EndGlobal
```

And this file can always be opened by Visual Studio 16.8.x to 16.9.x without any error, as the `MinimumVisualStudioVersion ` is still valid solution version.

If I use VS 2019 16.9.2 to create new blank solution, I get this:

```
Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio Version 16
VisualStudioVersion = 16.0.31112.23
MinimumVisualStudioVersion = 10.0.40219.1
Global
        GlobalSection(SolutionProperties) = preSolution
                HideSolutionNode = FALSE
        EndGlobalSection
        GlobalSection(ExtensibilityGlobals) = postSolution
                SolutionGuid = {195F37AF-9751-4154-9434-59DB85FFDC3F}
        EndGlobalSection
EndGlobal
```

Update that file by setting MinimumVisualStudioVersion to `16.0.31112.23`, like below:. 

```
Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio Version 16
VisualStudioVersion = 16.0.31112.23
MinimumVisualStudioVersion = 16.0.31112.23
Global
        GlobalSection(SolutionProperties) = preSolution
                HideSolutionNode = FALSE
        EndGlobalSection
        GlobalSection(ExtensibilityGlobals) = postSolution
                SolutionGuid = {195F37AF-9751-4154-9434-59DB85FFDC3F}
        EndGlobalSection
EndGlobal
```

Open this file in Visual Studio, and it will open nicely!

Notice the `VisualStudioVersion` generated by 16.9.2. It is versioned as `16.0.31112.23`, so VS will not see the version generated by .NET Core 5.0.201 SDK as valid version, because version `16.6.30114.105` is definitely **_higher_** than the current generated version of 16.9,2. 

## Suggestion

Please fix this solution creation version in .NET 5.0.201 SDK!

### PR DESCRIPTION
**Pull request by eriawan**
### Problem
I already detailed the problems in issue #2992

### Solution
Use valid solution version number for VS 2019 that always use this scheme of `16.0.xxxx.xxxx` instead of having invalid solution version number of `16.6.xxx.xx` as this breaks actual valid solution's minimum version. See also this comment: https://github.com/dotnet/templating/issues/2992#issuecomment-815281973
Please review. The detailed reasons why I submit this in issue #2992 

FYI, I'm using the version that reverted back to original solution version in the initial history of this file, based on this commit in this branch:  
https://github.com/dotnet/templating/commit/95d19ea6e0d10fc839f273223a736279464b3d38#diff-59d7a44760febde566cb20d6692d2a810281017b42d23841dd1e33eafa6275b1

image

### Checks:
- [ ] Added unit tests (_I think it is N/A as this is to fix to use correct solution's version number_)
- [ ] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/Solution/Solution1.sln
Comment (marcpopMSFT) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/Solution/Solution1.sln:4]: Is there a reason you went to 15 for VS Version rather than just fixing the 16 number to be 16.0.30114.105 which is the correct value for 16.6?

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/Solution/Solution1.sln
Comment (eriawan) [reply template_feed/Microsoft.DotNet.Common.ItemTemplates/content/Solution/Solution1.sln:5]: @marcpopMSFT 

Pardon, could you explain what is the technical reason of using version number of 10.0.xxx.xxx? What about the previous behavior in .NET Core 3.0 and 3.1 SDKs?
I mean, as far as I know, all of the solution generated by .NET Core 3.0.xxx, and also in Core 3.1.101 to 3.1.407 SDK always have minimum version set to `15.0.26124.0`, for example, this is the SLN generated from 3.1.406 SDK:

```
Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio 15
VisualStudioVersion = 15.0.26124.0
MinimumVisualStudioVersion = 15.0.26124.0
Global
        GlobalSection(SolutionConfigurationPlatforms) = preSolution
                Debug|Any CPU = Debug|Any CPU
                Debug|x64 = Debug|x64
                Debug|x86 = Debug|x86
                Release|Any CPU = Release|Any CPU
                Release|x64 = Release|x64
                Release|x86 = Release|x86
        EndGlobalSection
        GlobalSection(SolutionProperties) = preSolution
                HideSolutionNode = FALSE
        EndGlobalSection
EndGlobal
```

But if this change of versioning behavior from previous Core 3.0 and 3.1 is intended and should be enforced, then I will change this to use version `10.0.40219.1` instead. Please confirm and thanks! 🙂

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/Solution/Solution1.sln
Comment (eriawan) [reply template_feed/Microsoft.DotNet.Common.ItemTemplates/content/Solution/Solution1.sln]: Thanks for the suggestion! 
I have just updated with the version in this commit https://github.com/dotnet/templating/pull/3007/commits/a971eee1f63af668b7d83bec27ebf01aaf0c3575

**Comment by eriawan** [comment]
@marcpopMSFT 
I have to rebase it, because I mixed different git config between my local and Github direct edit.

Is there anything else besides your comment in https://github.com/dotnet/templating/pull/3007#discussion_r610119707 ?
Please confirm if this behavior changes are intended because the Core 3.0 and 3.1 always has `MinimumVisualStudioVersion ` set to other version higher than `10.0.xxxx.xxx`. when you are available.

Many thanks for your patience and review!

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/Solution/Solution1.sln
Comment (marcpopMSFT) [reply template_feed/Microsoft.DotNet.Common.ItemTemplates/content/Solution/Solution1.sln:5]: I'd try to have dotnet match what VS does which is target 10.  It's possible 3.0 picked 15 since .net core didn't exist back in 10.0.  Ultimately, I'll leave this to @donJoseLuis as the owner of the templating repo to make the call on what he wants the values to be.

Comment (eriawan) [reply template_feed/Microsoft.DotNet.Common.ItemTemplates/content/Solution/Solution1.sln]: @marcpopMSFT 

Ah, you're absolutely right! I have checked, and the `MinimumVisualStudioVersion` from VS 2019 is `10.0.40219.1`

I have updated this MinimumVisualStudioVersion` to use the 10.0.40219.1` version number.
Please review.

I also welcome @donJoseLuis to review 🙂

**Comment by eriawan** [comment]
@marcpopMSFT 

I think I have addressed all of your feedback. 

cc @donJoseLuis 

If you have some time, please review and thanks in advance, guys! 🙂 🙏

**Comment by eriawan** [comment]
hi @donJoseLuis 

Is there anything I should do further for this PR? Please review...

cc @marcpopMSFT

**Review by DavidKarlas — CHANGES_REQUESTED** [review]
Hi, could you undo BOM change at beginning of file? I'm worried VS or some other tool might break without it which some unicode project/file names.

**Review by dsplaisted — APPROVED** [review]

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/Solution/Solution1.sln
Comment (dsplaisted) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/Solution/Solution1.sln:1]: Apparently there are some whitespace changes here.  I don't know if that will affect anything but I think they should be reverted to be safe.

Comment (eriawan) [reply template_feed/Microsoft.DotNet.Common.ItemTemplates/content/Solution/Solution1.sln]: @dsplaisted 
Thanks!
Done, whitespace changes reverted.

**Comment by eriawan** [comment]
@DavidKarlas 

Thanks for the catch! My bad!
I have reverted to have correct BOM.  Please review, thanks in advance!

**Comment by eriawan** [comment]
@dsplaisted @DavidKarlas 

I think I have addressed all of the remaining PR feedback.
Please review and let me know if you have other feedback. 
Thanks in advance! 🙂 🙏

**Review by DavidKarlas — APPROVED** [review]

**Comment by eriawan** [comment]
@dsplaisted @DavidKarlas 

Pardon I keep bugging you, folks...
Now that this PR has been approved, is there anything else I have to do/edit? ❓ 
Because I'd love to port this PR to branches of `release/5.0.1xx` and `release\5.0.3xx` if this got merged. 🙏 

Thanks in advance!

**Comment by eriawan** [comment]
folks,

Please let me know, is this PR needs further triage/discussion for the team, or should I add more/update more? 
Since this PR has been approved, do you need more time to discuss with the rest of .NET Core team? Or more approvals?
Please confirm. It's been 3 days with no further update.. 

cc @dsplaisted @DavidKarlas

**Comment by DavidKarlas** [comment]
Hi @eriawan,

problem is that this PR is targeting `2xx` branch and not `3xx` or `main`, if you open PR against `3xx` and `main` we can merge that right away.

Merging things  into `2xx` is much harder. Since we need to argument why change meets bar for servicing and honestly, I don't think this bug is big enough to justify servicing and I wanted input from others before bringing it up.

I think that modifying .sln by hand is super edge case scenario and if you are advanced user enough to modify .sln by hand you can probably figure out what versions to use.

Sorry for not responding sooner.

Regards,
David

**Comment by eriawan** [comment]
@DavidKarlas 

Many thanks for your update!
About this:

> problem is that this PR is targeting 2xx branch and not 3xx or main, if you open PR against 3xx and main we can merge that right away.

Could I just directly change this PR's target branch to `release/5,0.2xx` to 'release/5.0.3xx` or do you want me to close this and create new 2 PR target 'release/5.0.3xx` branch and `main` branch separately?

Please confirm.

**Comment by DavidKarlas** [comment]
Preferably change this PR to `main` so once its merged, in future people can get back to this PR with all the context.
And open new one against `3xx`.

**Comment by eriawan** [comment]
@DavidKarlas 

done changing target branch to `main`, and now I have merge conflicts.
going to resolve these first...

**Comment by eriawan** [comment]
@DavidKarlas 

I have updated base branch to main as you have suggested, but then I got too many conflicting files...
since the conflicts are too many for me to quickly resolve, could I open a new PR with target branch to `main` instead?

**Comment by eriawan** [comment]
@DavidKarlas 

Are you currently busy now? I mean, please advice, as I have changed the target branch to `main` as suggested then I have conflicting files.

I still think it's easier to just create new PR that target `main` branch, instead of trying to resolve these numerous conflicting files. And yes, I agree that the context will not be lost if I just change the target branch. It's now I don't know how to qiuckly resolve these conflicting files from github web interface directly instead of resolving on my machine one by one.

Do you want me to solve this one by one for each conflicting files now? This will take some time, and I don't think I could finish this within 5 to 10 minutes, so I may revisit this later if I have enough time... 

Please let me know what do you think/suggest about this. Or, if you are busy, please let me know soon so I could revisit this PR in other time when you are available, while I continue to use 3.1.xxx instead for my team and my customers.

**Comment by DavidKarlas** [comment]
Its OK to close this PR and open new one, I forgot it would create all this conflicts, sorry about that...
We can just mention this PR in other PRs and it will be fine!

**Comment by eriawan** [comment]
@DavidKarlas 

Many thanks!
New PR to target `main` branch has been submitted: #3026 
I'm closing this PR now.

**Comment by DavidKarlas** [comment]
Can you also open for `3xx`?

**Comment by eriawan** [comment]
@DavidKarlas 

> Can you also open for 3xx?

Sure! I'm going to submit now

**Comment by eriawan** [comment]
I'm going to retarget this PR to 5.0.2xx in case this PR is approved to be included as servicing.

**Comment by eriawan** [comment]
Closing this PR after changing target branch back to 5.0.2xx.

**Comment by DavidKarlas** [comment]
Unfortunately this doesn't meet servicing bar but since its merged into 5.0.3xx branch, it will be available on stable channel in May which would also be true if it was merged into 5.0.2xx...

**Comment by eriawan** [comment]
@DavidKarlas 

Thanks for your reply!

> it will be available on stable channel in May which would also be true if it was merged into 5.0.2xx

So 5.0.3xx will be available on May? what do you mean "_**would also be true if it was merged into 5.0.2xx**_" ?

**Comment by DavidKarlas** [comment]
I meant to say that next 2xx version will also be released in May.

**Pull request closed without merging**
